src/java.base/share/classes/jdk/internal/misc/X-ScopedMemoryAccess.java.template

```diff
@@ -183,21 +184,56 @@ public class ScopedMemoryAccess {
     }
 
     @ForceInline
-    public void setMemory(MemorySessionImpl session, Object o, long offset, long bytes, byte value) {
+    public void setMemory(AbstractMemorySegmentImpl segment, byte value) {
         try {
-            setMemoryInternal(session, o, offset, bytes, value);
+            setMemoryInternal(segment, segment.byteSize(), value);
         } catch (ScopedAccessError ex) {
             throw ex.newRuntimeException();
         }
     }
 
+    // FILL_NATIVE_THRESHOLD must be a power of two and should be greater than 2^3
+    // Update the value for Aarch64 once 8338975 is fixed.
+    private static final long FILL_NATIVE_THRESHOLD = 1L << (Architecture.isAARCH64() ? 10 : 5);
+
     @ForceInline @Scoped
-    private void setMemoryInternal(MemorySessionImpl session, Object o, long offset, long bytes, byte value) {
+    private void setMemoryInternal(AbstractMemorySegmentImpl segment, long length, byte value) {
+        final MemorySessionImpl session = segment.sessionImpl();
         try {
             if (session != null) {
                 session.checkValidStateRaw();
             }
-            UNSAFE.setMemory(o, offset, bytes, value);
+            // 0...0X...XXXX implies: 0 <= length < FILL_NATIVE_LIMIT
```

Comment:
I would also recommend against "ultra" fine tuning and have limits that are different from platform to platform. It seems like we're using a different limit here because of https://bugs.openjdk.org/browse/JDK-8338975. But I'm not sure working that around is in the scope of the current PR.